staging/src/k8s.io/client-go/tools/cache/delta_fifo.go

```diff
@@ -163,8 +163,8 @@ func (f *DeltaFIFO) KeyOf(obj interface{}) (string, error) {
 // Return true if an Add/Update/Delete/AddIfNotPresent are called first,
 // or an Update called first but the first batch of items inserted by Replace() has been popped
 func (f *DeltaFIFO) HasSynced() bool {
-	f.lock.Lock()
-	defer f.lock.Unlock()
+	f.lock.RLock()
+	defer f.lock.RUnlock()
```

Comment:
My question is again the same - what does it give us? Have you seen any evidence this becoming a problem?

We should close this one.

Reply:
I feel the programming style is inconsistent: lock.RLock is used by DeltaFIFO#List but not DeltaFIFO#HasSynced

Reply:
Without read lock:

BenchmarkHasSynced-12                  	  200000	      7044 ns/op	    1046 B/op	      38 allocs/op
BenchmarkHasSynced-12                  	  200000	      6501 ns/op	    1046 B/op	      38 allocs/op
BenchmarkHasSynced-12                  	  200000	      6798 ns/op	    1046 B/op	      38 allocs/op
BenchmarkHasSynced-12                  	  200000	      6269 ns/op	    1046 B/op	      38 allocs/op
BenchmarkHasSynced-12                  	  200000	      6484 ns/op	    1046 B/op	      38 allocs/op

With read lock on HasSynced:

BenchmarkHasSynced-12                  	  200000	      6295 ns/op	    1046 B/op	      38 allocs/op
BenchmarkHasSynced-12                  	  200000	      6484 ns/op	    1046 B/op	      38 allocs/op
BenchmarkHasSynced-12                  	  200000	      6323 ns/op	    1046 B/op	      38 allocs/op

I would say read lock brought better performance.
Here is the benchmark:
```
func BenchmarkHasSynced(t *testing.B) {
	tests := []struct {
		actions        []func(f *DeltaFIFO)
		expectedSynced bool
	}{
		{
			actions:        []func(f *DeltaFIFO){},
			expectedSynced: false,
		},
		{
			actions: []func(f *DeltaFIFO){
				func(f *DeltaFIFO) { f.Add(mkFifoObj("a", 1)) },
				func(f *DeltaFIFO) { f.List() },
			},
			expectedSynced: true,
		},
		{
			actions: []func(f *DeltaFIFO){
				func(f *DeltaFIFO) { f.Replace([]interface{}{}, "0") },
				func(f *DeltaFIFO) { f.List() },
			},
			expectedSynced: true,
		},
		{
			actions: []func(f *DeltaFIFO){
				func(f *DeltaFIFO) { f.Replace([]interface{}{mkFifoObj("a", 1), mkFifoObj("b", 2)}, "0") },
				func(f *DeltaFIFO) { f.List() },
			},
			expectedSynced: false,
		},
		{
			actions: []func(f *DeltaFIFO){
				func(f *DeltaFIFO) { f.Replace([]interface{}{mkFifoObj("a", 1), mkFifoObj("b", 2)}, "0") },
				func(f *DeltaFIFO) { Pop(f) },
			},
			expectedSynced: false,
		},
		{
			actions: []func(f *DeltaFIFO){
				func(f *DeltaFIFO) { f.Replace([]interface{}{mkFifoObj("a", 1), mkFifoObj("b", 2)}, "0") },
				func(f *DeltaFIFO) { Pop(f) },
				func(f *DeltaFIFO) { Pop(f) },
			},
			expectedSynced: true,
		},
	}

	t.ReportAllocs()
	t.ResetTimer()
	f := NewDeltaFIFO(testFifoObjectKeyFunc, nil)
	for n := 1; n <= t.N; n++ {
		for _, test := range tests {
			cont := true
			go func(f *DeltaFIFO) {
				for cont == true {
					f.HasSynced()
				}
			}(f)
			for _, action := range test.actions {
				action(f)
			}
			cont = false
		}
	}
}
```

Reply:
I wouldn't say it's visibly better (e.g. the best run is actually for regular lock)

Also - HasSynced is being called really rarely, so it shouldn't matter at all...

Reply:
Looking at the median, performance with read lock was better.